experiments-core/src/foedus/ycsb/ycsb_driver.cpp

```diff
@@ -0,0 +1,374 @@
+      // again, numa_alloc_onnode should be better than numa_alloc_interleaved
+      options.memory_.interleave_numa_alloc_ = true;
+    }
+  }
+
+  options.snapshot_.folder_path_pattern_ = "/dev/shm/foedus_ycsb/snapshot/node_$NODE$";
+  options.log_.folder_path_pattern_ = "/dev/shm/foedus_ycsb/log/node_$NODE$/logger_$LOGGER$";
+  options.log_.loggers_per_node_ = FLAGS_loggers_per_node;
+  options.log_.flush_at_shutdown_ = false;
+  options.snapshot_.snapshot_interval_milliseconds_ = 100000000U;
+
+  options.debugging_.debug_log_min_threshold_
+    = debugging::DebuggingOptions::kDebugLogInfo;
+    // = debugging::DebuggingOptions::kDebugLogWarning;
+  options.debugging_.verbose_modules_ = "";
+  options.debugging_.verbose_log_level_ = -1;
+
+  options.log_.log_buffer_kb_ = FLAGS_log_buffer_mb << 10;
+  std::cout << "log_buffer_mb=" << FLAGS_log_buffer_mb << "MB per thread" << std::endl;
+  options.log_.log_file_size_mb_ = 1 << 15;
+  std::cout << "volatile_pool_size=" << FLAGS_volatile_pool_size << "GB per NUMA node" << std::endl;
+  options.memory_.page_pool_size_mb_per_node_ = (FLAGS_volatile_pool_size) << 10;
+
+  if (FLAGS_thread_per_node != 0) {
+    std::cout << "thread_per_node=" << FLAGS_thread_per_node << std::endl;
+    options.thread_.thread_count_per_group_ = FLAGS_thread_per_node;
+  }
+
+  // Get an engine, register procedures to run
+  Engine engine(options);
+  proc::ProcAndName load_proc("ycsb_load_task", ycsb_load_task);
+  proc::ProcAndName work_proc("ycsb_client_task", ycsb_client_task);
+  engine.get_proc_manager()->pre_register(load_proc);
+  engine.get_proc_manager()->pre_register(work_proc);
+  COERCE_ERROR(engine.initialize());
+  {
+    UninitializeGuard guard(&engine);
+    // This driver will fire off loading and impersonate clients
+    YcsbDriver driver(&engine);
+    driver.run();  // this will wait for the loader thread to finish
+    COERCE_ERROR(engine.uninitialize());
+  }
+  return 0;
+}
+
+ErrorStack YcsbDriver::run() {
+  // Setup the channel so I can synchronize with workers and record nr_workers
+  YcsbClientChannel* channel = get_channel(engine_);
+  channel->initialize();
+
+  // Figure out we have to run how many worker threads first.
+  // The loader needs it to generate keys.
+  const EngineOptions& options = engine_->get_options();
+  for (uint16_t node = 0; node < options.thread_.group_count_; node++) {
+    for (uint16_t ordinal = 0; ordinal < options.thread_.thread_count_per_group_; ordinal++) {
+      // XXX: this should be drawn from numa map actually?
+      channel->nr_workers_++;
+    }
+  }
+
+  auto* thread_pool = engine_->get_thread_pool();
+  thread::ImpersonateSession load_session;
+  bool ret = thread_pool->impersonate("ycsb_load_task", &channel->nr_workers_,
+                                      sizeof(uint32_t), &load_session);
+  if (!ret) {
+    LOG(FATAL) << "Couldn't impersonate";
+  }
+
+  // Wait for the load task to finish
+  // TODO(tzwang): parallelize this
```

Comment:
Yes, we need to do that. Not just about loading time, all volatile pages will be created at one socket. Log gleaner will also misunderstand it as data locality and place snapshot pages there.